### ISSUE TITLE
Sab 1.1.x - Copy's 1 of the par2's to the complete folder

### ISSUE THREAD

**Issue by mesa57** [sabnzbd/sabnzbd#724]
When option : 
- Download all par2 files is on
- Quickcheck is off
  And a repair is done, because 1 of the files outside of the rar is missing (nfo file), sabnzbd does not delete one of the .par2 (seems random which) and copies that to the complete folder.

**Comment by Safihre** [comment]
I don't completely get it? Can you give some more specific filenames of what's happening? 
Sometimes indeed if a par2 file is not used (didn't need the blocks of that file or was part of a different set), it gets copied info the final directory.

**Comment by mesa57** [comment]
If I set "download all par2 files" to on, then the download contains multiple, for instance 8, par2 files.
Only 1 is used, but the one copied to the complete folder is random and not always equal to the one used.
Seems to me that sabnzbd should remove all par2 files after succesfull repair.

**Comment by Safihre** [comment]
Could you send me an example NZB at safihre@sabnzbd.org?

**Comment by mesa57** [comment]
Send

**Comment by mesa57** [comment]
I did some analyses of the code and debug load. newsunpack.py does not register some file as processed when processing the output of par2.
Essential part of the debug log : 

```
Processing verifications and repairs concurrently using up to 2 logical CPUs.
Loading "155dbd33d75d0021a1680171916da4b3.par2".
Loaded 16 new packets
Loading "155dbd33d75d0021a1680171916da4b3.vol000+001.par2".
Loaded 1 new packets including 1 recovery blocks
Loading "155dbd33d75d0021a1680171916da4b3.vol001+002.par2".
Loaded 2 new packets including 2 recovery blocks
Loading "155dbd33d75d0021a1680171916da4b3.vol003+003.par2".
Loaded 3 new packets including 3 recovery blocks
Loading "155dbd33d75d0021a1680171916da4b3.vol006+005.par2".
Loaded 5 new packets including 5 recovery blocks
Loading "155dbd33d75d0021a1680171916da4b3.vol011+011.par2".
Loaded 11 new packets including 11 recovery blocks
Loading "155dbd33d75d0021a1680171916da4b3.vol022+022.par2".
Loading: 4.3%  <<<============ Loaded xx new packets is missing. This file is not registered
Loading "155dbd33d75d0021a1680171916da4b3.vol044+044.par2".
Loading: 4.5%
Loaded 22 new packets including 22 recovery blocks <<<========= Now the loaded is appearing
Loading "155dbd33d75d0021a1680171916da4b3.vol088+087.par2".
Loading: 13.7%
Loading: 37.8%
Loading: 63.1%
Loading: 87.3%
Loaded 87 new packets including 87 recovery blocks
Loading "155dbd33d75d0021a1680171916da4b3.vol175+173.par2".
Loading: 6.9%
Loading: 19.6%
Loading: 32.3%
Loading: 50.2%
Loaded 44 new packets including 44 recovery blocks
Loading: 68.7%
Loading: 82.6%
Loading: 94.7%
Loaded 173 new packets including 173 recovery blocks
There are 7 recoverable files and 0 other files.
The block size used was 224216 bytes.
There are a total of 2326 data blocks.
The total size of the data files is 521290163 bytes.
Verifying source files:
Scanning: "155dbd33d75d0021a1680171916da4b3.r00": 0.2%
```

As a consequence file 155dbd33d75d0021a1680171916da4b3.vol022+022.par2 is copied to the output.

**Comment by mesa57** [comment]
Unchecking "use multicore par2" leads to miserable results. De set is verified OK and par2 does not even recognize that 1 file is missing ! Also ALL par2 files are subsequently copied to the complete folder. :(

So checked back "use multicore par2" to ON but also added the -t- parameter to force it to use a single core. Essential part of the debug log is then :

```
Processing verifications and repairs serially.
Loading "155dbd33d75d0021a1680171916da4b3.par2".
Loaded 16 new packets
Loading "155dbd33d75d0021a1680171916da4b3.vol000+001.par2".
Loaded 1 new packets including 1 recovery blocks
Loading "155dbd33d75d0021a1680171916da4b3.vol001+002.par2".
Loaded 2 new packets including 2 recovery blocks
Loading "155dbd33d75d0021a1680171916da4b3.vol003+003.par2".
Loaded 3 new packets including 3 recovery blocks
Loading "155dbd33d75d0021a1680171916da4b3.vol006+005.par2".
Loaded 5 new packets including 5 recovery blocks
Loading "155dbd33d75d0021a1680171916da4b3.vol011+011.par2".
Loaded 11 new packets including 11 recovery blocks
Loading "155dbd33d75d0021a1680171916da4b3.vol022+022.par2".
Loading: 72.4%
Loaded 22 new packets including 22 recovery blocks
Loading "155dbd33d75d0021a1680171916da4b3.vol044+044.par2".
Loading: 63.5%
Loaded 44 new packets including 44 recovery blocks
Loading "155dbd33d75d0021a1680171916da4b3.vol088+087.par2".
Loading: 33.2%
Loading: 94.2%
Loaded 87 new packets including 87 recovery blocks
Loading "155dbd33d75d0021a1680171916da4b3.vol175+173.par2".
Loading: 26.5%
Loading: 57.7%
Loading: 89.0%
Loaded 173 new packets including 173 recovery blocks
There are 7 recoverable files and 0 other files.
The block size used was 224216 bytes.
There are a total of 2326 data blocks.
The total size of the data files is 521290163 bytes.
Verifying source files:
Scanning: "155dbd33d75d0021a1680171916da4b3.r00": 0.2%
```
After that all par2 files are recognized by sabnzbd and subsequently deleted. 
It appears this is a at least a workaround :smile: 

But I think the multicore issue should be fixed.

**Comment by Safihre** [comment]
Yes this is exactly true and inspecting the code this seems to be exactly the desired functioning.
I think (was written a while ago) it's because this way we only delete par2 files that actually belong to the set.
It is for example possible that we tell par2 to scan a specific file (we tell it to scan the par2's we think belong to a set on the command line) that actually belongs to a different set. Or maybe it's some special ```rename.par2``` that can later be used  to rename the outcome file.
So we don't just want to delete all par2, only when we are very sure that it belongs to the set we are currently repairing and repair was sucessful. 
I guess you can say we are over-protective..
But you can also just put ```par2``` in the Config->Switches->Cleanup list.

**Comment by mesa57** [comment]
It is fine that you only delete par2 files actually belonging to the set.
However, in a multicore par2 execution, the algoritme which checks this does not work !
As you can see in the first debug log snippet, the order of messages appearing from multicore par2 is not sequentially and therefore the code to determine if the par2 belongs to the set does not work correct.

**Comment by Safihre** [comment]
Ah yes I should have been more clear:
In case a file does not belong to the set and we tell par2 to scan it while it's not part of the set, the output would look the same.
To illustrate I just added a random file to a folder, renamed it to end on ```.vol003+003.par2``` and ran verify on ```809908887080907.par2```.
This is the output:
```
Loading "809908887080907.par2".
Loaded 36 new packets
Loading "809908887080907.vol000+001.par2".
Loaded 1 new packets including 1 recovery blocks
Loading "809908887080907.vol003+003.par2". <======== This is a fake file, nothing is reported
Loading "809908887080907.vol006+005.par2".
Loaded 5 new packets including 5 recovery blocks
etc etc
```
So we don't know if it was part of the set, this is all it tells us. Just nothing.

**Comment by mesa57** [comment]
I know what you're code is trying to do, matching "loaded xx packets" with previous lines with "Loading {par2 filename}" and only accepting when a match is found.
However, multicore PAR2 messes this up and randomly par files are not accepted.
The NZB which I send you as example does not contain any par files not belonging to the set. 
But sabnzbd thinks so, because it cannot match one of those randomly.
This becomes more apparent when you check "download all par files" to ON, because then more things are executed in parallel. But it could also easily occur when more then `1 par file is needed for repair.

**Comment by mesa57** [comment]
In case of multicore par2 I think the only thing which you can do, is not try to match the Loading and Loaded, but just count them. And if the count is unequal, just delete no par files at all.
But that is a suggestion of my side, maybe there are better solutions.

**Comment by Safihre** [comment]
Aah, I see now what you mean. There's indeed overlapping output! 
But knowing the code, what could we do about it possibly? Disable multicore just to not have maybe 1 or 2 leftover files?

What I can think of is to delete all par2 files that we detected (during download) when all repairing is done and successfull. This should be possible.

**Comment by mesa57** [comment]
What about my suggestion ? I think it is a (little) bit more safer then just deleting them all.
An of course only under the condition that par2 is running multicore. Detectable by this line : 

`Processing verifications and repairs concurrently using up to 2 logical CPUs`

**Comment by Safihre** [comment]
So it turns out all that was necessary was a simple change, since there's already code that detects all the par2 files of a set.
See bbf7964
All we have to do is just add to the list all par2 files of the set and remove those as well, on top of any par2 files that ```par2``` detected extra.
